OSDOCS-10463: vsphere multiple subnets (#75717)
openshift/openshift-docs: modules/nw-operator-vsphere-multiple-subnets.adoc

// Module included in the following assemblies:
//
// * installing/installing_vsphere/ipi/installing-vsphere-installer-provisioned-network-customizations.adoc
// * installing/installing_vsphere/upi/installing-vsphere-network-customizations.adoc

:_mod-docs-content-type: PROCEDURE
[id="nw-operator-vsphere-multiple-subnets_{context}"]
= Specifying multiple subnets for your network

> Before you install an {product-title} cluster on a {vmw-short} host, you can specify multiple subnets for a networking implementation so that the {vmw-short} cloud controller manager (CCM) can select the appropriate subnet for a given networking situation. The {vmw-full} can use the subnet for managing pods and services on your cluster. For this configuration, you must specify internal and external Classless Inter-Domain Routing (CIDR) implementations in the {vmw-short} cloud controller manager (CCM) configuration. Each CIDR implementation lists an IP address range that the CCM uses to decide what subnets interact with traffic from internal and external networks.

Reviewer: * `The {vmw-full}` renders as "The VMware vSphere" which seems like it should just be "vSphere"?
* Suggest breaking this long paragraph up a bit for easier reading :)
* "CCM" already defined earlier. I'm actually not sure we should abbreviate it, upstream doesn't tend to but other docs sites are a mixed bag.

```suggestion
Before you install an {product-title} cluster on a {vmw-short} host, you can specify multiple subnets for a networking implementation so that the {vmw-short} cloud controller manager (CCM) can select the appropriate subnet for a given networking situation. {vmw-short} can use the subnet for managing pods and services on your cluster. 

For this configuration, you must specify internal and external Classless Inter-Domain Routing (CIDR) implementations in the {vmw-short} CCM configuration. Each CIDR implementation lists an IP address range that the CCM uses to decide what subnets interact with traffic from internal and external networks.
```

Author: :heavy_check_mark:

> If you do not configure internal and external CIDR implementations in the {vmw-short} CCM configuratio, the {vmw-short} CCM might select the wrong subnet. This configuration can cause new nodes that associate with a `MachineSet` object with a single subnet to become unusable as each new node receives the `node.cloudprovider.kubernetes.io/uninitialized` taint. These situations can cause communication issues with the Kubernetes API server that can cause installation of the cluster to fail.

Reviewer 2: typo missing `n` in `configuratio,`

```
in the {vmw-short} CCM configuratio,
```

Reviewer 2: other than typo, LGTM.

> If you do not configure internal and external CIDR implementations in the {vmw-short} CCM configuration, the {vmw-short} CCM might select the wrong subnet. This configuration can cause new nodes that associate with a `MachineSet` object with a single subnet to become unusable as each new node receives the `node.cloudprovider.kubernetes.io/uninitialized` taint. These situations can cause communication issues with the Kubernetes API server that can cause installation of the cluster to fail.
> 
> .Example error message because of the {vmw-short} CCM selecting an incompatible subnet
> [source, text]
> ----
> ERROR Bootstrap failed to complete: timed out waiting for the condition
> ERROR Failed to wait for bootstrapping to complete. This error usually happens when there is a problem with control plane hosts that prevents the control plane operators from creating the control plane.
> ----

Reviewer: This seems important enough for some sort of admonition, and would suggest getting the error block up closer to the part of the failure that it references. Something like this (feel free to tweak):

```suggestion
[IMPORTANT]
====
Failure to configure internal and external CIDR implementations in the {vmw-short} CCM configuration can cause the {vmw-short} CCM to select the wrong subnet. This causes the following error:

.Example error message
[source, text]
----
ERROR Bootstrap failed to complete: timed out waiting for the condition
ERROR Failed to wait for bootstrapping to complete. This error usually happens when there is a problem with control plane hosts that prevents the control plane operators from creating the control plane.
----

This configuration can cause new nodes that associate with a `MachineSet` object with a single subnet to become unusable as each new node receives the `node.cloudprovider.kubernetes.io/uninitialized` taint. These situations can cause communication issues with the Kubernetes API server that can cause installation of the cluster to fail.
====
```

Author: :heavy_check_mark:

.Prerequisites

* You created Kubernetes manifest files for your {product-title} cluster.

.Procedure

. From the directory where you store your {product-title} cluster manifest files, open the `manifests/cluster-infrastructure-02-config.yml` manifest file.

. Add a `nodeNetworking` object to the file and specify internal and external network subnet CIDR implementations for the object.
+

> [NOTE]

Reviewer: Maybe a TIP? (I might be biased because I think that admonition type is neat, feel free to ignore)
```suggestion
[TIP]
```

// Module included in the following assemblies:
//
// * installing/installing_vsphere/ipi/installing-vsphere-installer-provisioned-network-customizations.adoc
// * installing/installing_vsphere/upi/installing-vsphere-network-customizations.adoc

:_mod-docs-content-type: PROCEDURE
[id="nw-operator-vsphere-multiple-subnets_{context}"]
= Specifying multiple subnets for your network

Before you install an {product-title} cluster on a {vmw-short} host, you can specify multiple subnets for a networking implementation so that the {vmw-short} cloud controller manager (CCM) can select the appropriate subnet for a given networking situation. The {vmw-full} can use the subnet for managing pods and services on your cluster. For this configuration, you must specify internal and external Classless Inter-Domain Routing (CIDR) implementations in the {vmw-short} cloud controller manager (CCM) configuration. Each CIDR implementation lists an IP address range that the CCM uses to decide what subnets interact with traffic from internal and external networks.

> If you do not configure internal and external CIDR implementations in the {vmw-short} CCM configuration, the {vmw-short} CCM selects the first subnet that the host operating system reports as the primary network interface. This configuration can cause new nodes that associate with a `MachineSet` object with a single subnet to become unusable as each new node receives the `node.cloudprovider.kubernetes.io/uninitialized` taint. In certain situations, the reported network interface might not be the expected network interface. These situations can cause communication issues with the Kubernetes API server that can cause installation of the cluster to fail.

Reviewer 3: This is a little weird now because I'm not sure what the behavior is in a node with a single NIC and multiple subnets. Does the CCM just pick the first subnet? I would assume so, but I don't know.

Maybe we just drop the specific explanation of how it works and say that if you do not configure these values the wrong subnet may be selected and then keep the discussion of the uninitialized taint? Maybe remove the "In certain situations, the reported network interface might not be the expected network interface." line too since it doesn't entirely make sense in a single NIC scenario either, and I think the important points about the uninitialized taint causing deployment failure are still there.

[IMPORTANT]
====

> To prevent any IP address selection issues for each IP protocol, only configure the amount of IP addresses and IP protocols that you will use. If you configured multiple subnets for a network implementation but you intend to only use the IP protocols on a single subnet, ensure that you deploy your cluster nodes on the single subnet.

Reviewer 3: As we discussed on Slack, I think we can remove this now. I think it's unlikely people are configuring extra subnets, and I don't want to confuse things by talking about NICs in a section that's mostly focused on subnets.
